### PR TITLE
Added CUDA graph, Tensor Core and Core pinning explaination

### LINKED ISSUE #2861: Performance Tuning Guide is very out of date
### 🚀 Descirbe the improvement or the new tutorial

The first thing you see when you Google PyTorch performance is this. The recipe is well written but it's very much out of data today
https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html

Some concrete things we should fix
1. For fusions we should talk about torch.compile instead of jit.script
2. We should mention overhead reduction with cudagraphs
3. We should talk about the *-fast series as places people can learn more
4. For CPU specific optimization the most important one is launcher core pinning so we should either make that a default or explain the point more
5. Instead of the CPU section we can instead go more into the inductor CPU backend
6. AMP section is fine but maybe expand to quantization
7. DDP section needs to be moved somewhere else with some FSDP performance guide
8. GPU sync section is good
9. Mention tensor cores and how to enable them and why they're not enabled by default

cc @sekyondaMeta @svekars @kit1980 @drisspg who first made me aware of this with an internal note that was important enough to make public

### Existing tutorials on this topic

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by orion160**
Fixes #2861

cc @sekyondaMeta @svekars @kit1980 @brycebortree

### CODE REVIEW

File: recipes_source/recipes/tuning_guide.py [hunk 1]
```diff
@@ -318,6 +319,32 @@ def gelu(x):
...
+# operations which neural network operation can take advantage of.
+#
+# Hardware tensor core operations tend to use a different floating point format
+# which sacrifices precision at expense of speed gains.
+# Prior to pytorch 1.12 this was enabled by default but since this version
+# it must be explicitly set as it can conflict with some operations which do not
+# benefit from Tensor core computations.
+
+
+torch.backends.cuda.matmul.allow_tf32
```
Comment (msaroufim) [inline recipes_source/recipes/tuning_guide.py:335]: favor `torch.set_float32_matmul_precision("high")` instead

File: recipes_source/recipes/tuning_guide.py [hunk 2]
```diff
@@ -318,6 +319,32 @@ def gelu(x):
...
+
+###############################################################################
+# Use CUDA Graphs
+# ~~~~~~~~~~~~~~~~~~~~~~~
+# At the time of using a GPU, work first must be launched from the CPU and
+# on some cases the context switch between CPU and GPU can lead to bad resourse
+# utilization. CUDA graphs are a way to keep computation within the GPU without
+# paying the extra cost of kernel launches and host synchronization.
+#
+# It can be enabled using `torch.compile <https://pytorch.org/docs/stable/generated/torch.compile.html>`_ "reduce-overhead" and "max-autotune" modes.
```
Comment (msaroufim) [inline recipes_source/recipes/tuning_guide.py:345]: can be more explicit and say `torch.compile(m, mode="reduce-overhead")` to save people a click

**Comment by orion160** [comment]
@msaroufim resolved

**Review by msaroufim — APPROVED** [review]

**Comment by orion160** [comment]
Hmmmm it seems that the spell checker doesnt like

```
# and will set torch.backends.cuda.matmul.allow_tf32 = True if available

**Comment by msaroufim** [comment]
Id just get rid of the line

File: recipes_source/recipes/tuning_guide.py [hunk 3]
```diff
@@ -318,6 +319,42 @@ def gelu(x):
 # GPU specific optimizations
 # --------------------------
 
+###############################################################################
+# Enable Tensor cores
+# ~~~~~~~~~~~~~~~~~~~~~~~
+# Tensor cores are specialized hardware to compute matrix-matrix multiplication
```
Comment (svekars) [inline recipes_source/recipes/tuning_guide.py:325]:
```suggestion
# Tensor cores are specialized hardware designed to compute matrix-matrix multiplication
```

File: recipes_source/recipes/tuning_guide.py [hunk 4]
```diff
@@ -318,6 +319,42 @@ def gelu(x):
...
+# operations which neural network operation can take advantage of.
```
Comment (svekars) [inline recipes_source/recipes/tuning_guide.py:326]:
```suggestion
# operations, which neural network operations can take advantage of.
```

File: recipes_source/recipes/tuning_guide.py [hunk 5]
```diff
@@ -318,6 +319,42 @@ def gelu(x):
...
+#
+# Hardware tensor core operations tend to use a different floating point format
+# which sacrifices precision at expense of speed gains.
+# Prior to pytorch 1.12 this was enabled by default but since this version
```
Comment (svekars) [inline recipes_source/recipes/tuning_guide.py:330]: since which version? 2.3? @msaroufim 
```suggestion
# Prior to PyTorch 1.12 this functionality was enabled by default but since this version
```

File: recipes_source/recipes/tuning_guide.py [hunk 6]
```diff
@@ -318,6 +319,42 @@ def gelu(x):
...
+# it must be explicitly set as it can conflict with some operations which do not
+# benefit from Tensor core computations.
+
+## Tensor computation can be enabled "manually" modifying the matrix multiplication precision
+## The default precision is "highest" which will perform the operation according to the dtype
```
Comment (svekars) [inline recipes_source/recipes/tuning_guide.py:335]: I'm not sure why it has ##, did you intend it to be rendered as a code block? I think it would make sense for this text to be a regular para? You can check rendering [here](https://docs-preview.pytorch.org/pytorch/tutorials/2912/recipes/recipes/tuning_guide.html#enable-tensor-cores).

File: recipes_source/recipes/tuning_guide.py [hunk 7]
```diff
@@ -318,6 +319,42 @@ def gelu(x):
...
+# precision "high" and "medium" can be hardware accelerated via tensor cores
+
+# Carefully consider the tradeoff between speed and precision at the moment of evaluating your models!
+torch.set_float32_matmul_precision("high")
+
+###############################################################################
+# Use CUDA Graphs
+# ~~~~~~~~~~~~~~~~~~~~~~~
+# At the time of using a GPU, work first must be launched from the CPU and
+# on some cases the context switch between CPU and GPU can lead to bad resource
```
Comment (svekars) [inline recipes_source/recipes/tuning_guide.py:346]:
```suggestion
# in some cases the context switch between CPU and GPU can lead to bad resource
```

File: recipes_source/recipes/tuning_guide.py [hunk 8]
```diff
@@ -318,6 +319,42 @@ def gelu(x):
...
+# utilization. CUDA graphs are a way to keep computation within the GPU without
+# paying the extra cost of kernel launches and host synchronization.
+
+# It can be enabled using 
+torch.compile(m, "reduce-overhead")
+# or
+torch.compile(m, "max-autotune")
+
+###############################################################################
+# Special care must be present when using cuda graphs as it can lead to increased memory consumption and some models might not compile.
```
Comment (svekars) [inline recipes_source/recipes/tuning_guide.py:356]: I wasn't sure what is "it" in it can lead to increased memory, so suggesting "incorrect implementation" - feel free to modify as needed: 
```suggestion
# Special care must be taken when using CUDA graphs, as incorrect implementation can lead to increased memory consumption and some models might not compile.
```

**Review by svekars — COMMENTED** [review]
Thanks so much for the updates. I added a few suggestions. Also, you can check rendering [here](https://docs-preview.pytorch.org/pytorch/tutorials/2912/recipes/recipes/tuning_guide.html#enable-tensor-cores).

**Comment by orion160** [comment]
@svekars Thanks for the comments!

**Review by svekars — APPROVED** [review]
